### DISCUSSION TITLE
Why is calling `sync` from a running loop a bad idea?

### DISCUSSION THREAD

**Discussion by freddyaboulton** [fsspec/filesystem_spec#1570] [Q&A] [answered]
Hi all!

I just got this error: https://github.com/fsspec/filesystem_spec/blob/05e7d80ab2affaa01505ff2602b0097c38ad7688/fsspec/asyn.py#L80

I'm wondering why that exception is there in the first place? Thank you

**Comment by martindurant** [comment, marked as answer]
Unfortunately, the event loop in asyncio is non-reentrant. That means, you cannot say "hey loop, run this" from within the loop. The only way to do that is `await`, but you can't do that in a non-async function. So you must do one or the other: use `sync()`, or `await` on async methods.

`sync()` works by using a dedicated IO thread, since we *can* wait on thread signaling primitives live `event` in a normal function. If you are running your whole application as async, you don't need another thread.

**Reply by martindurant** [reply]
Wee also https://github.com/erdewit/nest_asyncio , which was one attempt to solve the problem, but a low-level library like fsspec shouldn't alter the runtime for all other possible users.

**Reply by freddyaboulton** [reply]
Thank you for the explanation @martindurant !

**Comment by MarcSkovMadsen** [comment]
I also go this error when using fsspec in [HoloViz Panel](). I'll probably have to switch to async interface.

**Reply by martindurant** [reply]
Sorry, I don't think fsspec can call nest_syncio, since that changes session-wide definitions of builtin behaviour - so you are stuck with choosing the right API.

**Reply by MarcSkovMadsen** [reply]
Figured out the problem is with adfls and not fsspec as such. See https://github.com/holoviz/panel/issues/6934#issue-2370107143.
